pkg/tls/cert_loader_test.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2020 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package tls
+
+import (
+	"crypto/tls"
+	"io/ioutil"
+	"os"
+	"testing"
+	"time"
+)
+
+func Test_PeriodicReload(t *testing.T) {
+	rootCAFile, err := ioutil.TempFile("", "rootCA.crt")
+	if err != nil {
+		t.Fatalf("error creating tmp root CA: %v", err)
+	}
+	defer os.Remove(rootCAFile.Name())
+	if _, err := rootCAFile.Write([]byte(testRootCA)); err != nil {
+		t.Fatalf("error writing to tmp root CA: %v", err)
+	}
+
+	serverCertFile, err := ioutil.TempFile("", "server.crt")
+	if err != nil {
+		t.Fatalf("error creating tmp server cert: %v", err)
+	}
+	defer os.Remove(serverCertFile.Name())
+	if _, err := serverCertFile.Write([]byte(testServerCert)); err != nil {
+		t.Fatalf("error writing to tmp server cert: %v", err)
+	}
+
+	serverKeyFile, err := ioutil.TempFile("", "server.key")
+	if err != nil {
+		t.Fatalf("error creating tmp server key: %v", err)
+	}
+	defer os.Remove(serverKeyFile.Name())
+	if _, err := serverKeyFile.Write([]byte(testServerKey)); err != nil {
+		t.Fatalf("error writing to tmp server key: %v", err)
+	}
+
+	certLoader, err := NewPeriodicCertLoader(serverCertFile.Name(), serverKeyFile.Name(), 1*time.Second)
+	if err != nil {
+		t.Fatalf("error starting periodic cert reloader: %v", err)
+	}
+
+	serverConfig := &tls.Config{}
+	serverConfig.GetCertificate = func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+		return certLoader.Current(), nil
+	}
+	clientConfig := &tls.Config{}
+	clientConfig.RootCAs = prepareCertPool(rootCAFile.Name())
+
+	// first check client -> server w/ TLS to validate initial reload worked
+	testClientServerHello(t, clientConfig, serverConfig)
+
+	// start certificate loader and nil out the current certificate, should expect an error
+	go certLoader.Start()
+	certLoader.current = nil
+	testClientServerHelloError(t, clientConfig, serverConfig)
```

Comment:
Do we need a delay before running this test? Also, shouldn't we be setting`certLoader.current` to `nil` prior to starting the cert loader?

Reply:
> Do we need a delay before running this test? 

Yes this is to test that the certificate is reloaded after the sync period.  

> Also, shouldn't we be setting certLoader.current to nil prior to starting the cert loader?

I don't think so because we're testing whether the certificate is reloaded from disk after the sync period.

Reply:
I meant I don't see a delay before `testClientServerHelloError` is called, so was wondering if we're checking the certs loaded by the `certLoader` at all in this test.

Reply:
Ah I see. There's an initial reload as part of `NewPeriodicCertLoader`, which is how `testClientServerHello` on line 55 is passing. I agree that it would be more intuitive to start the periodic sync after we check the error case. I'll update to do this

Reply:
Updated, hope it reads better now :)

Reply:
Looks better now, thanks.